### ISSUE TITLE
[5.2 & 5.3] Compiling C++ code using the OCaml C API results in name-mangled caml_state on Cygwin

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/ocaml#13541]
Trying to compile opam (which uses C++ through mccs) on Cygwin with the 5.3 branch at 162a7f4869bb3136b48fd811408cc6f53396a0ea, fails with:
```
  ** Cannot resolve symbols for descriptor object:
   _ZTH10caml_state
  File "caml_startup", line 1:
  Error: Error during linking (exit code 2)
```
Whereas it compiles just fine with OCaml 4.14.2 with the same environment.

**Comment by MisterDA** [comment]
Is it just missing in `caml/domain_state.h` an `extern "C"` block to prevent C++ name mangling?

```c
#ifdef __cplusplus
extern "C" {
#endif

/* … */

#ifdef __cplusplus
}
#endif
```

**Comment by kit-ty-kate** [comment]
probably yes

**Comment by MisterDA** [comment]
I could not reproduce.
I built an OCaml 4.14 compiler using the 4.14 branch and the Cygwin port, which I used to build opam 2.2.1, then built a OCaml 5.3 compiler (d750a53ae69bcf302c37fe514f5781cfe85d5bd5) with the Cygwin port, created an opam switch out of it, loaded the switch and checked that `ocamlc -config-var os_type` showed `Cygwin`, then downloaded and successfully built mccs 1.1+18.
Perhaps I've done something wrong, but if you could provide a reproducer, show where the error appeared, or the invocation that failed (with `dune build --verbose`), that could help.

**Comment by kit-ty-kate** [comment]
building mccs is not enough as it is a library. The error is a linking error. You can reproduce it by going slightly further and trying to install the `opam-devel` package which creates a binary at the end.

**Comment by MisterDA** [comment]
Indeed, I can reproduce by running the tests.

**Comment by kit-ty-kate** [comment]
could this issue also be added to the 5.3.0 milestone?

**Comment by xavierleroy** [comment]
Yes, let's fix this for 5.3.  But the problem has been there since 5.0, right?

**Comment by kit-ty-kate** [comment]
5.0 didn't have cygwin support. 5.1 and 5.2 had it but i just tried with 5.2.0 and it indeed also fails the same way.
My bad, i thought it was a 5.3 regression, but it's only a bug since 5.1 or 5.2.

**Comment by kit-ty-kate** [comment]
actually i just tried 5.1.1 and it worked fine. So this is a 5.2 regression

**Comment by kit-ty-kate** [comment]
Could this be reopened? This is broken again with OCaml 5.4.0~beta1

**Comment by gasche** [comment]
This is addressed in #13777, right? (@MisterDA, you mentioned that you wanted to split this issue, what is the current state?) Is this something that possibly @dra27 would be able to review?

**Comment by Octachron** [comment]
@kit-ty-kate , is this a new failure in the beta or was the issue potentially here during the first alpha?

**Comment by kit-ty-kate** [comment]
I just tried with 5.4.0~alpha1 and it also failed

**Comment by dra27** [comment]
Just FTR, the regression is in fact #12811 (in 5.2.0), and the use of `thread_local` for `caml_state` in C++ mode - see https://github.com/ocaml/ocaml/pull/14220#issuecomment-3254629771
